jiant/preprocess.py

```diff
@@ -614,3 +634,17 @@ def add_wsj_vocab(vocab, data_dir, namespace="tokens"):
     for line in wsj_tokens.readlines():
         vocab.add_token_to_namespace(line.strip(), namespace)
     log.info("\tAdded WSJ vocabulary from %s", wsj_tokens)
+
+
+class TaskModulator(object):
```

Comment:
In a comment of #845, we have discussed about boundary_token_fn, how did a model-specific thing winded up in task. Now, when we add task more pytorch_transformers models, whether task.process_split should connect the sentence pair into one big sequence is depend on which input_module we are using, it will be too ugly if we import some function defined by a particular model, then use it in task.py, since task related code would better be model-independent.
I think there is a slightly more systematic way. If we look back on the big picture, the rendezvous between task designers and model designers is most commonly strings of text, sometimes with other fields like numbers, task designers won't specify how the text should be preprocessed (except wt103), it is the model designers that decide which tokenizer to use(especially so for large pretrained models), whether to solely rely on character-level etc. Task is mostly static, unchanged through out the program running. While model is dynamic, it needs to be computed over and over. Yet, one part decided by the model has the characteristic of being static, the preprocessing, including tokenization, boundary_token_fn, whether to combine sentence pair etc.
Ideally, preprocess receive the necessary (and nothing more than that) information / objects from model, form a preprocess function from them, and process the task data with it, before the model starts running. Under this view, the "create-copies-of-BERT-tokenizer-everywhere" problem (see the TODO in pytorch_transformer_interface/modules.py) also become clear -- the only real BERT tokenizer belongs to the model, when we use it to tokenize text, or use its vocab to index tokens, it is only because model passes it to preprocess. Even the default vocab created from scratch belong to model, the only difference being vocab from scratch forms a more sophisticated preprocess function.
(BTW, the counterpart of preprocess, something decided by task designers, but is dynamic, also exist, that is the task metrics. In this view, functions about how to update metrics should be created by tasks, then passed to the model, and running dynamically, I think jiant already employed this idea)
Surely, this small, shallow class of "TaskModulator" does not fully embody this view, but it is a step in that direct while still coherent to the current framework.

Reply:
I see the problem, and I think this change is reasonable, but I don't completely understand how this is improving things now. Is there some duplication that's going away?

Reply:
Yes, marginally. Instead of deciding whether the model support pair embedding every time we call process_split, we do it only once here in preprocess. But comparing to the time cost of entire preprocess, the saving is negligible. The main benefit lies in extensibility. If there is a new model that process input differently on some tasks, we only need to add a new flag(or reuse flag when possible) here, and modify the process_split in related tasks. In comparison, our old method would require adding a new arguments to every process_split function in the code.

Reply:
Okay. I'm not completely convinced that this is helpful, but I don't think it's harmful, so I'm okay merging it like this.

Reply:
Thanks! I believe this is a good way to organize it. Take the example of "mirrored_pair", this is a special mechanism that only works when model is gpt/gpt2/transfo-xl and task is MRPC/STS-B, it is due to this module that we can implement this with minimal change in tasks.py . I'm totally fine if you have different opinion, I just want to make sure what I'm thinking is clearly expression.

Reply:
Works for me!